### PR TITLE
Enforce Strict `@types` Dependencies

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

As part of my work on #37280, I will be centralizing our TypeScript configuration to make maintenance easier. In the process of doing that, however, I found that there were some `@types` declarations missing from a few packages. By default, TypeScript will search for types in `./node_modules` and will traverse parent directories looking for the same. This can cause issues when a `@types` declaration is missing from the package but a parent directory has one from a different version of the package. 

This pull request isolates the types to the project's directory, and as a consequence, caused a number of type errors across the monorepo. This pull request contains fixes for all of those errors.

In the interesting of an easier time reviewing and testing, this pull request does not contain any TypeScript refactoring changes. The changes in this PR are **just** the result of fixing any errors introduced by setting `"typeRoots"` in all of the `tsconfig.json` and `tsconfig-cjs.json` files.

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

* Verify that all of the CI checks pass.
* Verify that the output of `pnpm run test --filter="!woocommerce"` matches what was present before this PR.
* Verify that WooCommerce's React admin pages still work as expected.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [x] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/components/src/utils.tsx [hunk 1]
```diff
@@ -64,7 +64,7 @@ function createOrderedChildren< T = Fill.Props, S = Record< string, unknown > >(
 	order: number,
 	props: T,
 	injectProps?: S
-) {
+): React.ReactSVGElement {
```
Comment (ilyasfoo) [inline packages/js/components/src/utils.tsx]: I think this is unintended? `React.ReactElement` seems to be the better suited type
```suggestion
): React.ReactElement {
```

File: packages/js/product-editor/src/utils/create-ordered-children.tsx [hunk 2]
```diff
@@ -64,7 +64,7 @@ function createOrderedChildren< T = Fill.Props, S = Record< string, unknown > >(
 	order: number,
 	props: T,
 	injectProps?: S
-) {
+): React.ReactSVGElement {
```
Comment (ilyasfoo) [inline packages/js/product-editor/src/utils/create-ordered-children.tsx]: Same as above

```suggestion
): React.ReactElement {
```

**Review by chihsuan — DISMISSED** [review]
Awesome! Looks good to me and tested well. 👍

**Review by ilyasfoo — CHANGES_REQUESTED** [review]
Thanks for working on this, @ObliviousHarmony! This tested well, I have a minor comment:

File: packages/js/tsconfig-cjs.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
 	"extends": "./tsconfig",
 	"compilerOptions": {
-		"target": "es6",
+		"target": "esnext",
```
Comment (louwie17) [inline packages/js/tsconfig-cjs.json]: This is quite the jump, not that I am opposed. But will this cause any issues for the newest release of packages, for the plugins that bundle it within their code?
We did probably have to create major release versions for each new package or not?

Comment (louwie17) [reply packages/js/tsconfig-cjs.json]: Although it may be fine anyhow given the `module` hasn't changed - my `tsconfig` knowledge is quite limited.

**Review by louwie17 — COMMENTED** [review]
Looks good from my end, aside from the two items that @ilyasfoo has already pointed out around the `ReactElement` type.

File: packages/js/components/src/utils.tsx (hunk 1 above)
Comment (ObliviousHarmony) [reply packages/js/components/src/utils.tsx]: I went with what VS Code told me that the function would be returning 😄 Thanks for the tip here, I'll make the change!

File: packages/js/tsconfig-cjs.json (hunk 3 above)
Comment (ObliviousHarmony) [reply packages/js/tsconfig-cjs.json]: Ultimately the goal of the `target` option is to tell TypeScript what features the output JavaScript should be able to make use of. On the browser front, Gutenberg's `tsconfig` files also target `"esnext"` and their Babel configuration targets browsers that WordPress has declared compatibility for.

With that said, however, I think you're right that this is a bit heavy-handed. I'll change this to `"es2019"` which is the same as our `tsconfig.json`. This is supported by the version of Node we've declared compatibility with, as well as all of the browsers that we are compatible with. Given that our ESM builds already target this I don't think it requires a major version release.

**Review by ilyasfoo — DISMISSED** [review]
LGTM! 🚢

**Review by ilyasfoo — APPROVED** [review]
Changes lgtm and tests well! 🚢

**Pull request merged**